external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/SoftwareBf16.cpp

```diff
@@ -0,0 +1,364 @@
+//===- LegalizeForExport.cpp - Prepare for translation to LLVM IR ---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/LLVMIR/Transforms/SoftwareBF16.h"
+#include "PassDetail.h"
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+namespace mlir {
+namespace LLVM {
+/// Rewrites bf16 constants to their i16 equivalents
+/// This is relying on the fact that the vector, i16, and bf16 types used in the
+/// LLVM dialect are the standard ones and not weird custom wrappers
+struct BF16ConstCasting : OpRewritePattern<LLVM::ConstantOp> {
+  explicit BF16ConstCasting(MLIRContext *context) : OpRewritePattern(context) {}
+
+  llvm::APInt toInt(llvm::APFloat value) const {
```

Comment:
Name: `castBf16ToInt` or such.

Also, make this a `static` helper outside the struct so [the people I've mainly been working with] upstream don't complain about why you didn't do that.